front/src/applications/operationalStudies/helpers/upsetNewProjectedTrains.ts

```diff
@@ -0,0 +1,39 @@
+import type { TrainSpaceTimeData } from 'applications/operationalStudies/types';
+import { type ProjectPathTrainResult, type TrainScheduleResult } from 'common/api/osrdEditoastApi';
+import { convertIsoUtcToLocalTime } from 'utils/date';
+import type { Dictionary } from 'utils/types';
+
+// TODO: feed the spaceTimeChart with date instead of formatted string
+const formatDatetimeForSpaceTimeChart = (departureTime: string) =>
+  `${convertIsoUtcToLocalTime(departureTime).slice(0, -6)}Z`;
+
+const upsertNewProjectedTrains = (
+  projectedTrains: Dictionary<TrainSpaceTimeData> | undefined,
+  projectedTrainsToUpsert: Dictionary<ProjectPathTrainResult>,
+  trainSchedulesById: Dictionary<TrainScheduleResult>
+) => {
+  const newProjectedTrains = { ...projectedTrains };
+
+  // For each key (train id) in projectPathTrainResult, we either add it or update it in the state
+  Object.keys(projectedTrainsToUpsert).forEach((trainId) => {
+    const currentProjectedTrain = projectedTrainsToUpsert[trainId];
+    if (!currentProjectedTrain) {
+      console.error(`Train ${trainId} not found in the projectedTrainsToUpsert`);
+      return;
+    }
+
+    const matchingTrain = trainSchedulesById[trainId];
+    const projectedTrain = {
+      ...currentProjectedTrain,
+      id: +trainId,
+      trainName: matchingTrain?.train_name || 'Train name not found',
+      departure_time: formatDatetimeForSpaceTimeChart(currentProjectedTrain.departure_time),
+    };
+
+    newProjectedTrains[trainId] = projectedTrain;
+  });
+
+  return newProjectedTrains;
+};
```

Comment:
We can improve performance a little bit by using` Object.entries` here instead of `Object.keys` , as it provides both keys and values directly, making the code a bit cleaner and avoiding extra lookups.

```suggestion
  Object.entries(projectedTrainsToUpsert).forEach(([trainId, trainData]) => {
    if (!trainData) {
      console.error(`Train ${trainId} not found in the projectedTrainsToUpsert`);
      return;
    }

    const schedule = trainSchedulesById[trainId];
    const projectedTrain = {
      ...trainData,
      id: Number(trainId),
      trainName: schedule?.train_name || 'Train name not found',
      departure_time: formatDatetimeForSpaceTimeChart(currentProjectedTrain.departure_time),
    };

    newProjectedTrains[trainId] = projectedTrain;
  });

  return newProjectedTrains;
};
```

Reply:
Done